### PR TITLE
Fixup null reference access in Breadcrumbs `computeVisibleItems`.

### LINKED ISSUE #1979: Uncaught TypeError: Cannot read property 'children' of null from Breadcrumbs.tsx
# 🐛 Bug Report 

Seeing a traceback in the console when using `Breadcrumbs` inside a `Dialog`. 

This is a little tricky to reproduce, but my current hunch is it has to do with the Dialog the Breadcrumbs component is a child of being unmounted while Breadcrumbs is busy working on (checks notes) computing its visible items.

Part of the JSX, hopefully the relevant parts, looks like:

```jsx
<DialogContainer onDismiss={close}>
      {isOpen && (
        <Dialog size="L">
          <Heading data-testid="menu-heading">
            <Breadcrumbs onAction={(key) => transitionTo(key)}>
              {parents.map((node) => (
                <Item key={node.key}>{node.value.name}</Item>
              ))}
            </Breadcrumbs>
          </Heading>
....
``` 

Anecdotally, it seems as though when this `DialogContainer` is a child of another component that unmounts, we see the error.

## 🤔 Expected Behavior

I'd hope the `Breadcrumbs` component would not throw this error.

## 😯 Current Behavior

```
Uncaught TypeError: Cannot read property 'children' of null
    at computeVisibleItems (Breadcrumbs.tsx:59)
    at Breadcrumbs.tsx:112
    at Generator.next (<anonymous>)
    at Object.nextRef.current (useValueEffect.ts:31)
    at Object.nextRef.current (useValueEffect.ts:43)
    at useValueEffect.ts:58
    at Breadcrumbs.tsx:107
    at ResizeObserver.<anonymous> (useResizeObserver.ts:33)
```

## 💁 Possible Solution

If I had to guess, the code in the ResizeObserver is running after unmount, sort of a "use after free" situation. Not sure how best to mitigate.

## 💻 Code Sample

I have not been able to build a small targeted repro case. Sorry!

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   | 3.10.0 |
| Browser          | Chrome 91.0.4472.77 |
| Operating System | Linux |

### PR DESCRIPTION
**Pull request by onelson**
TL;DR the `current` field on a ref is `T | null` not `T` and we must
guard against invalid access by first interrogating `current`.

Since refs are initialized as null and can become null at runtime, blind
access to `current` can lead `TypeError`s (as cited in #1979).

In practice it seems this _can happen reliably_ in situations where a
`Breadcrumb` component is being unmounted as part of a larger component
tree.

Unfortunately, I have not been able to reduce this error case to a small
reproduction, therefore I have not been able to add a test explicitly
to signal regressions.

Still, and this is quite troubling, the Typescript compiler should be
flagging invalid access such as this. I'm not sure if something has been
disarmed in the tsconfig to assume `T | null` is `T` in this case, but
a similar access in a [TS playground](https://www.typescriptlang.org/play?ts=3.8.3#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AoczAVwDsNgJbElMBJWzJI+pACgCUcAN7k4cMXDRMAzvCKY4AXhboYAOmoykyTAB4AEgBUAsgBkAqmeByAogBskIJLRgA+PrWr37AiuMkAekCWTBkpFGYAIyQ4Lx84FHk6GFAkdUlpWjkpAAtgewATHmU4BXU0aigeDTR8op4KSSIYKuZ4+woAX3IgA)
is flagged as an error.

This aims to fix #1979.

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

**Comment by onelson** [comment]
Anecdotally, this does appear to squash the runtime errors I saw in my app.

I understand the best case solution would be to somehow ensure the observer callback is unregistered earlier (ref: https://github.com/adobe/react-spectrum/issues/1979#issuecomment-879460558), but short-circuiting `computeVisibleItems` here will go a long way to ease the development of the apps we are working on day-to-day.

As such, you can consider this a bandaid.

---

As far as the whole type-check story is concerned, we've been running into similar problems elsewhere in the Spectrum ecosystem where signatures indicate a `T` but at runtime we see either `undefined` or `null` in some cases. The specific case I can recall right now is in the `TreeData#getItem()` method. It would be much better for us as consumers of these APIs if the signatures represented the uncertainty we actually see at runtime.

Perhaps this is a separate topic worth a separate issue.

File: packages/@react-spectrum/breadcrumbs/src/Breadcrumbs.tsx [hunk 1]
```diff
@@ -56,8 +56,14 @@ function Breadcrumbs<T>(props: SpectrumBreadcrumbsProps<T>, ref: DOMRef) {
 
   let updateOverflow = useCallback(() => {
     let computeVisibleItems = (visibleItems: number) => {
-      let listItems = Array.from(listRef.current.children) as HTMLLIElement[];
-      let containerWidth = listRef.current.offsetWidth;
+      // N.B. refs can be null at runtime.
```
Comment (snowystinger) [inline packages/@react-spectrum/breadcrumbs/src/Breadcrumbs.tsx:59]: odd question, but what is N. B.?

**Review by snowystinger — COMMENTED** [review]
Thank you! and yeah, we have some things to sort out with regards to TS :)

File: packages/@react-spectrum/breadcrumbs/src/Breadcrumbs.tsx (hunk 1 above)
Comment (onelson) [reply packages/@react-spectrum/breadcrumbs/src/Breadcrumbs.tsx:59]: _Nota bene_, meaning "observe carefully" :nerd_face:

Comment (onelson) [reply packages/@react-spectrum/breadcrumbs/src/Breadcrumbs.tsx:59]: FWIW, I think once TS is flagging this sort of access, this would not be worthy of a callout. The compiler would make it self-evident.

Happy to rephrase or remove the comment if the larger concern of nulls/undefined is already well known among the maintainers here.

Comment (snowystinger) [reply packages/@react-spectrum/breadcrumbs/src/Breadcrumbs.tsx:59]: I'm not worried about calling it out, but maybe since we have a fairly diverse set of people who need to be able to read the code, we can just leave off the N B, the fact that there is a comment is probably enough to say "warning here"

**Review by snowystinger — APPROVED** [review]
Thanks again!

**Review by LFDanLu — APPROVED** [review]
LGTM, thanks for the fix!
